resolver/resolver.go

```diff
@@ -405,12 +410,40 @@ func isUDP(w dns.ResponseWriter) bool {
 	return strings.HasPrefix(w.RemoteAddr().Network(), "udp")
 }
 
-// truncate sets the TC bit in the given dns.Msg if its length exceeds the
-// permitted length of the given transmission channel.
+// truncate removes answers until the given dns.Msg fits the permitted
+// length of the given transmission channel and sets the TC bit.
 // See https://tools.ietf.org/html/rfc1035#section-4.2.1
-func truncate(m *dns.Msg, udp bool) *dns.Msg {
-	m.Truncated = udp && m.Len() > dns.MinMsgSize
-	return m
+func truncate(m *dns.Msg, udp bool) (*dns.Msg, error) {
+	max := dns.MinMsgSize
+	if !udp {
+		max = dns.MaxMsgSize
+	} else if opt := m.IsEdns0(); opt != nil {
+		max = int(opt.UDPSize())
+	}
+
+	m.Truncated = m.Len() > max
+	if !m.Truncated {
+		return m, nil
+	}
+
+	m.Extra = nil // Drop all extra records first
+
+	answers := make([]dns.RR, len(m.Answer))
```

Comment:
you shouldn't need to copy the slice contents here, instead create a snapshot of the slice (cap,len) that refers to the same underlying storage:
```go
answers := m.Answer[:]
```